src/plugins/memdump/userhook.cpp

```diff
@@ -0,0 +1,671 @@
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    bool our_fault = false;
+
+    auto vec_it = plugin->loaded_dlls.find(info->regs->cr3);
+
+    if (vec_it != plugin->loaded_dlls.end())
+    {
+        for (auto const& dll_meta : vec_it->second)
+        {
+            if (dll_meta.dtb == info->regs->cr3 && dll_meta.thread_id == thread_id)
+            {
+                our_fault = true;
+                break;
+            }
+        }
+    }
+
+    if (!our_fault)
+    {
+        PRINT_DEBUG("[MEMDUMP-USER] Not suppressing service exception - not our fault\n");
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    // emulate `ret` instruction
+    addr_t saved_rip = 0;
+    access_context_t ctx =
+    {
+        .translate_mechanism = VMI_TM_PROCESS_DTB,
+        .dtb = info->regs->cr3,
+        .addr = info->regs->rsp,
+    };
+
+    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+    bool success = (VMI_SUCCESS == vmi_read(vmi, &ctx, sizeof(addr_t), &saved_rip, NULL));
+    drakvuf_release_vmi(drakvuf);
+
+    if ( !success )
+    {
+        PRINT_DEBUG("[MEMDUMP-USER] Error while reading the saved RIP in system service handler\n");
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    constexpr int EXCEPTION_CONTINUE_EXECUTION = 0;
+    info->regs->rip = saved_rip;
+    info->regs->rsp += sizeof(addr_t);
+    info->regs->rax = EXCEPTION_CONTINUE_EXECUTION;
+    return VMI_EVENT_RESPONSE_SET_REGISTERS;
+}
+
+/**
+ * Observe process exit and remove all user mode hooks
+ */
+static event_response_t terminate_process_hook_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    auto plugin = get_trap_plugin<memdump>(info);
+    if (!plugin)
+        return VMI_EVENT_RESPONSE_NONE;
+
+    auto vec_it = plugin->loaded_dlls.find(info->regs->cr3);
+
+    if (vec_it == plugin->loaded_dlls.end())
+        return VMI_EVENT_RESPONSE_NONE;
+
+    for (auto& it : vec_it->second)
+    {
+        for (auto& target : it.targets)
+        {
+            PRINT_DEBUG("[MEMDUMP-USER] Erased trap for pid %d %s\n", info->proc_data.pid, target.target_name.c_str());
+            drakvuf_remove_trap(drakvuf, target.trap, NULL);
+        }
+    }
+
+    plugin->loaded_dlls.erase(vec_it);
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+void memdump::userhook_init(drakvuf_t drakvuf, const memdump_config* c, output_format_t output)
+{
+    breakpoint_in_system_process_searcher bp;
+    if (!register_trap<memdump>(drakvuf, nullptr, this, protect_virtual_memory_hook_cb, bp.for_syscall_name("NtProtectVirtualMemory")) ||
+        !register_trap<memdump>(drakvuf, nullptr, this, map_view_of_section_hook_cb, bp.for_syscall_name("NtMapViewOfSection")) ||
+        !register_trap<memdump>(drakvuf, nullptr, this, system_service_handler_hook_cb, bp.for_syscall_name("KiSystemServiceHandler")) ||
+        !register_trap<memdump>(drakvuf, nullptr, this, terminate_process_hook_cb, bp.for_syscall_name("NtTerminateProcess")))
```

Comment:
What happens with the traps that were successfully added when say the last one failed to register?

Reply:
There will be `throw -1;` and it will reach the `memdump::memdump` constructor, which calls `userhook_init` internally.

The same mode of operation is already in `procmon`: https://github.com/tklengyel/drakvuf/blob/72ba1f485477c4b5665e8deb33c990f15a3f2fbf/src/plugins/procmon/procmon.cpp#L721

Reply:
Well, it's perhaps incorrect in `procmon` as well :) I'm just wondering whether the traps that were added successfully are not removed after the `throw -1` happens.

Reply:
Normally there should be a linked list of traps that the plugin is actively using, and when it exits it goes through that list to delete the traps. I don't see the `plugin_ex` version of `register_trap` doing that.

Reply:
Doesn't it halt the whole DRAKVUF when a plugin startup fails to start?

Reply:
https://github.com/tklengyel/drakvuf/blob/master/src/plugins/plugins.cpp#L335

If any plugin would throw an exception in constructor then the whole DRAKVUF halts.

Reply:
It should yes, and `libdrakvuf` will ultimately clean up any registered traps, but still, I prefer that the plugins do that cleanup of their own traps in their destructor.

Reply:
See https://github.com/tklengyel/drakvuf/blob/master/src/plugins/syscalls/syscalls.cpp#L828 for an example

Reply:
Though its a problem of `plugins_ex` which is rather outside of this PR.

Reply:
I propose this to be a separate topic, as it requires changes at least few plugins and this PR is rather huge already.